src/datasources/accounts/notifications/notifications.datasource.ts

```diff
@@ -0,0 +1,234 @@
+
+@Injectable()
+export class NotificationsDatasource implements INotificationsDatasource {
+  constructor(
+    @Inject('DB_INSTANCE')
+    private readonly sql: postgres.Sql,
+    @Inject(LoggingService)
+    private readonly loggingService: ILoggingService,
+    @Inject(IAccountsDatasource)
+    private readonly accountsDatasource: IAccountsDatasource,
+  ) {}
+
+  /**
+   * Upserts subscriptions for the given account/device as per the list of Safes
+   * and notification types provided.
+   *
+   * @param args.account Account address
+   * @param args.cloudMessagingToken Cloud messaging token
+   * @param args.deviceType Device type
+   * @param args.deviceUuid Device UUID (defaults to random UUID)
+   * @param args.safes List of Safes with notification types
+   *
+   * @returns Device UUID
+   */
+  async upsertSubscriptions(
+    args: UpsertSubscriptionsDto,
+  ): Promise<{ deviceUuid: Uuid }> {
+    const account = await this.accountsDatasource.getAccount(args.account);
+    const deviceUuid = args.deviceUuid ?? crypto.randomUUID();
+
+    await this.sql.begin(async (sql) => {
+      // Get the push notifications channel
+      const [channel] = await sql<[{ id: number }]>`
+        SELECT id FROM notification_channels
+        WHERE name = ${DomainNotificationChannel.PUSH_NOTIFICATIONS}
+      `.catch((e) => {
+        const error = 'Error getting channel';
+        this.loggingService.info(`${error}: ${asError(e).message}`);
+        throw new NotFoundException(error);
+      });
+
+      // Insert (or update the cloud messaging token of) a device
+      const [device] = await sql<[{ id: number }]>`
+        INSERT INTO notification_devices (device_type, device_uuid, cloud_messaging_token)
+        VALUES (${args.deviceType}, ${deviceUuid}, ${args.cloudMessagingToken})
+        ON CONFLICT (device_uuid)
+        DO UPDATE SET
+          cloud_messaging_token = EXCLUDED.cloud_messaging_token,
+          -- Throws if updated_at is not set
+          updated_at = NOW()
+        RETURNING id
+      `.catch((e) => {
+        const error = 'Error getting device';
+        this.loggingService.info(`${error}: ${asError(e).message}`);
+        throw new UnprocessableEntityException(error);
+      });
+
+      // For each Safe, upsert the subscription and overwrite the subscribed-to notification types
+      await Promise.all(
+        args.safes.map(async (safe) => {
+          try {
+            // 1. Upsert subscription
+            const [subscription] = await sql<[{ id: number }]>`
+              INSERT INTO notification_subscriptions (account_id, chain_id, safe_address, device_id, notification_channel_id)
+              VALUES (${account.id}, ${safe.chainId}, ${safe.address}, ${device.id}, ${channel.id})
+              ON CONFLICT (account_id, chain_id, safe_address, device_id, notification_channel_id)
+                -- A field must be set to return the id
+                DO UPDATE SET updated_at = NOW()
```

Comment:
As before regarding the `updated_at` field.

Reply:
As [above](https://github.com/safe-global/safe-client-gateway/pull/1756#discussion_r1689629831).

Reply:
Updated the comment in 19900061b57560d9dde756cb76babc0a2493cd43.